<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-177349182-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-177349182-1');
    </script>

    <title>Home</title>
  </head>

  <body>
    {% extends "base.html" %}
    {% block jumbo %}
    <div class="container-fluid dark-background very-dark-text" style="padding:20px;">
        <div class="jumbotron light-pink-background">
          <div class="row">
            <div class="col-lg-2 col-md-2 col-sm-2 col-xs-2">
              <img src="static/img/Tree-transparent.png" width="90%;" height="85%;">
            </div>
            <div class="col-lg-8 col-md-8 col-sm-8 col-xs-8 ">
              <h1 class="display-4" style="text-align:center;">KinCoRe</h1>
              <p class="lead" style="text-align:center;">Kinase Conformation Resource</p>
              <hr class="my-4">
              <p class="very-dark-text" style="text-align:center;">A web resource for protein kinase sequence, structure and phylogeny.</p>
            </div>
           <div class="col-lg-2 col-md-2 col-sm-2 col-xs-2">
              <img src="static/img/EGFR-superposition-transparent.png" width="90%;" height="85%;" >
            </div>
          </div>
        </div>
<div class="jumbotron" style="background-color:white;">
          <div class="row">
            <div class="col-lg-12 col-md-12 col-sm-12 col-xs-12">
              <div class="row">
                <div class="col-lg-4 col-md-12 col-sm-12 col-xs-12">
                  <div class="jumbotron-fluid"  style="margin:25px;text-align:center;">
                    <h6><h4 style="display:inline;">Scan</h4> all the human structures in one table</h6>
                    <hr class="my-8 brand-background" style="border-width: 2px;">

                    <a class="btn button-primary" style="color:white;" href="{{ url_for('browse') }}" target="_blank" role="button">Browse</a>
                  </div>
                </div>

                <div class="col-lg-4 col-md-12 col-sm-12 col-xs-12">
                  <div class="jumbotron-fluid" style="margin:25px;text-align:center;">
                    <h6><h4 style="display:inline;">Query</h4> using an identifier, group or conformation</h6>
                    <hr class="my-8 brand-background" style="border-width: 2px;">
                    <h6></h6>
                    <a class="btn button-primary" href="{{ url_for('formSearch') }}" target="_blank" role="button">Search</a>
                  </div>
                </div>


                <div class="col-lg-4 col-md-12 col-sm-12 col-xs-12">
                  <div class="jumbotron-fluid"  style="margin:25px;text-align:center;">
                    <h6><h4 style="display:inline;">Upload</h4> a structure to determine its conformation</h6>
                    <hr class="my-8 brand-background" style="border-width: 2px;">

                    <a class="btn button-primary" href="{{ url_for('webserver') }}" target="_blank" role="button">Upload</a>
                  </div>
                </div>
              </div>

              <div class="row">
                <div class="col-lg-4 col-md-12 col-sm-12 col-xs-12 offset-lg-1">
                  <div class="jumbotron-fluid"  style="margin:25px;text-align:center;">
                    <h6><h4 style="display:inline;">Multiple</h4> sequence alignment of 497 human kinases</h6>
                    <hr class="my-8 brand-background" style="border-width: 2px;">

                    <a class="btn button-primary" href="{{ url_for('alignment') }}" target="_blank" role="button">Alignment</a>
                  </div>
                </div>

                <div class="col-lg-4 col-md-12 col-sm-12 col-xs-12 offset-lg-2">
                  <div class="jumbotron-fluid"  style="margin:25px;text-align:center;">
                    <h6><h4 style="display:inline;">Phylogenetic</h4> tree based on structural information</h6>
                    <hr class="my-8 brand-background" style="border-width: 2px;">

                    <a class="btn button-primary" href="{{ url_for('phylogeny') }}" target="_blank" role="button">Phylogeny</a>
                  </div>
                </div>
              </div>
            </div>
          </div>
</div>

        </div>
    </div>
    {% endblock %}

    {% block content %}

        <div class="row">
          <div class="col-lg-11 col-md-11 col-sm-11 col-xs-11">
            <div class="container-fluid very-dark-text">
              <div class="card mb-3">
                <div class="card-header light-pink-background">
                  <h5>Background</h5>
                </div>
                <div class="row no-gutters">
                  <div class="col-lg-6 col-md-12 col-sm-12 col-xs-12">
                    <div class="card-body">
                        <h5>Structure</h5>
                        <p class="card-text" style="text-align:justify;">Protein kinases (PKs) are enzymes that transfer phosphoryl group from an ATP molecule to Ser, Thr or Tyr residue of the substrate protein. The human genome consists 484 PK genes (497 domains) that are divided broadly into nine families based on their sequences namely, AGC, CAMK, CK1, CMGC, NEK, RGC, STE, TKL, TYR and OTHER (unclassified). They share a conserved structural fold consisting of two lobes: an N-terminal lobe, formed by five stranded β-sheet with an α-helix called the C-helix, and a C-terminal lobe comprising six α-helices. The two lobes are connected by a flexible region in the middle which forms the ATP binding active site of the protein.</p>
                        <br>
                        <h5>Activation loop</h5>
                        <p class="card-text" style="text-align:justify;">The activation loop is typically 20-30 residues in length and is the most critical secondary structural element of the active site of PKs. It is in completely extended conformation in the catalytically active state of the enzyme facilitating the binding of ATP molecule and the substrate. However, it folds on the surface of the protein in different kinds of inactive states. The activation loop begins with a conserved sequence motif called DFGmotif (Asp, Phe, and Gly residues). These residues are observed to be in a unique orientation when the loop is extended (active state) but display remarkable flexibility in folded (inactive) loop conformations.</p>
                    </div>
                  </div>
                  <div class="col-lg-3 col-md-6 col-sm-6 col-xs-6">
                    <br><br>
                    <img src="static/img/INSR_structure.jpg"  height="350px;">
                    <p style="text-align:center;">Typical structure of protein kinase</p>
                  </div>
                  <div class="col-lg-3 col-md-6 col-sm-6 col-xs-6">
                    <br><br>
                    <img src="static/img/EGFR-superposition-figure.png" width="350px;" height="350px;">
                    <p style="text-align:center;">Multiple orientations of DFG-Phe in EGFR</p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="row">
          <div class="col-lg-11 col-md-11 col-sm-11 col-xs-11">
            <div class="container-fluid very-dark-text">
              <div class="card mb-3">
                <div class="card-header light-pink-background">
                  <h5>Classification and nomenclature for active and inactive protein kinase structures</h5>
                </div>
                <div class="no-gutters">
                    <div class="card-body">
                      <div class="row">
                        <div class="col-lg-12 col-md-12 col-sm-12 col-xs-12">

                          <h5>We have clustered all the conformations of PK activation loop in two steps:<br><br>Spatial groups</h5>
                          <p style="text-align:justify;">We have determined the location of DFG-Phe ring in the binding pocket based on its distance from two
                            conserved residues,<br>
                            <ul>
                            <li>
                            D1 = dist(αC-Glu(+4)-Cα, DFG-Phe-Cζ) - distance between the Cα atom of the fourth residue after the conserved Glu residue
                            in the C-helix (ExxxX) and the outermost atom of the DFG-Phe ring (Cζ). This distance serves to
                            distinguish DFGin structures, where the Phe ring is adjacent to or under the C-helix, from DFGout structures, where the ring has
                            moved a substantial distance laterally from the C-helix.</li><br>
                            <li>D2= dist(β3-Lys-Cα, DFG-Phe-Cζ) - distance between the Cα atom
                            of the conserved Lys residue from the β3 strand to the Cζ atom of the DFG-Phe side chain. It captures the closeness of DFG-Phe to
                            the N-lobe β-sheet strands, thus giving an estimate of the upward position of the Phe ring, distinguishing DFGin conformations from
                            structures where the Phe ring is in an intermediate position between DFGin and DFGout.</li></p>
                          </ul>
                          <img src="{{ url_for('static', filename="img/Distance-plot.jpg") }}" width="685px;" height="310px;" align="right" style="padding:10px;">
                          <p style="text-align:justify;">Based on the spatial location of DFG-Phe ring in the binding pocket we have classified kinase structures into three broad groups:<br>
                            <ol>
                              <li>DFGin:  This is the largest group representing the DFG motif orientations where DFG-Phe is packed against or under the C-helix.
                            It consists of many related conformations with the typical DFGin active orientation forming the largest subset of this group.</li><br>
                              <li>DFGout: This is the second largest group representing the structures where DFG-Phe is moved into the ATP binding pocket. The
                            structures with a Type 2 inhibitor bound form a subset of this group.</li><br>
                              <li>DFGinter -  (DFGintermediate): This is the smallest group in which the DFG-Phe side chain is out of the C-helix pocket but has
                            not moved completely to a DFGout conformation. In most of these cases DFG-Phe is pointing upwards towards the β-sheets dividing the
                            active site into two halves.</li></p>

                      </div>
                    </div>
                  </div>
                </div>


              <hr class="my-4 brand-background" style="margin:20px;">
                <div class="no-gutters">
                    <div class="card-body">
                      <div class="row">
                        <div class="col-lg-6 col-md-12 col-sm-12 col-xs-12">
                          <h5>Dihedral clusters</h5>
                          <p style="text-align:justify;">Each spatial group consists of multiple closely related conformations. To cluster these conformations we used the backbone dihedrals
                            (φ,ψ) of X-DFG (residue before conserved Asp), DFG-Asp, DFG-Phe and side chain dihedral (χ1) of DFG-Phe. These dihedrals we used to
                            compute a distance matrix which is then provided as an input to DBSCAN (Density-based spatial clustering of applications with noise), a
                            density-based clustering algorithm. The different clusters observed are labeled on the basis of <a href="{{ url_for('static', filename="img/Rama-ABLE.png") }}" target="_blank">Ramachandran region (A, B, L, and E)</a>
                            occupied by XDF residue backbone and the DFG-Phe χ1 rotamer (minus = -60°; plus = +60°; trans = 180°).</p>
                        </div>
                        <div class="col-lg-6 col-md-12 col-sm-12 col-xs-12">
                          <img src="{{ url_for('static', filename="img/Dbscan.jpg") }}" alt="" width=675px>
                        </div>
                        <div class="col-lg-6 col-md-12 col-sm-12 col-xs-12">
                          <ul>
                          <li><p style="text-align:justify;">For the DFGin group we obtained six clusters labeled as BLAminus, BLAplus, ABAminus, BLBminus, BLBplus, and BLBtrans.
                            All the catalytically primed structures (ATP+Mg bound and activation loop phosphorylated) are observed in the BLAminus cluster.</li></p>
                          <li><p style="text-align:justify;">For the DFGout group we obtained just one cluster. In this cluster, the X-D-F residues occupy the B-B-A regions of the Ramachandran map and DFG-Phe is
                            in a -60° rotamer. More than 82% of Type 2 inhibitor bound structures are BBAminus; the remainder are in the DFGout noise group.</li></p>
                          <li><p style="text-align:justify;">The structures in the DFGinter conformation display more variability than the other states. For the DFGinter group we obtained only
                            one small cluster. The X-D-F residues are in a B-A-B conformation and the DFG-Phe residue is observed in a trans rotamer with a few chains
                            displaying a rotamer orientation between g-minus and trans.</li></p>
                          </ul>
                        </div>

                        <div class="col-lg-6 col-md-12 col-sm-12 col-xs-12">
                          <img src="{{ url_for('static', filename="img/Cluster-examples.jpg") }}" alt="" width=675px>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>

          <div class="row">
            <div class="col-lg-11 col-md-11 col-sm-11 col-xs-11">
              <div class="container-fluid very-dark-text">
                <div class="card mb-3">
                  <div class="card-header light-pink-background">
                    <a name="comparison_states"></a><h5>Comparisons of different conformational states</h5>
                  </div>


                  <div class="no-gutters">
                      <div class="card-body">
                        <div class="row">
                          <div class="col-lg-12">
                            <div class="row">
                            <div class="col-lg-6 col-md-12 col-sm-12 col-xs-12">

                            <p style="text-align:justify;"><b>A</b>. Pairwise comparisons of 5 different states of human BRAF kinase <br>DFGin-BLBplus structures (“SRC-inactive” conformation in blue with grayish-blue activation loop)
compared with DFGin-BLAminus (orange, active conformations), BLBminus (magenta-pink), DFGinBLAplus (“FGFR-inactive”, cyan-lightcyan), and DFGout-BBAminus (Type-2 ligand binding, lightbrownyellow). A small number of outlier structures in some classes are not shown. In each case, the position of the activation loop and the C-helix is correlated with
the state of the kinase, and differs among the 5 states.</p>Download in Pymol:
<ul>
<li><a href="{{ url_for('static',filename='img/Fig_A1_BRAF_BLBplus_blue_BLAminus_orange.pse.zip') }}" target="_blank">BRAF_BLBplus_BLAminus</a><br></li>
<li><a href="{{ url_for('static',filename='img/Fig_A2_BRAF_BLBplus_blue_BLBminus_cyan.pse.zip') }}" target="_blank">BRAF_BLBplus_BLBminus</a></li>
<li><a href="{{ url_for('static',filename='img/Fig_A3_BRAF_BLBplus_blue_BLAplus_green.pse.zip') }}" target="_blank">BRAF_BLBplus_BLAplus</a></li>
<li><a href="{{ url_for('static',filename='img/Fig_A4_BRAF_BLBplus_blue_BBAminus_yellow.pse.zip') }}" target="_blank">BRAF_BLBplus_BBAminus</a></li>
</ul>
<p style="text-align:justify;"><b>B</b>. BLAminus and BLBplus states of EGFR<br> EGFR BLAminus (107
chains, left, orange) and BLBplus (79 chains, right, blue) structures with both the activation loop
(right side of each image) and C-terminal tail (left side of each image) shown in lighter colors. The C-terminus of each group is shown in magenta spheres. The state of the activation loop is highly correlated with the
position of the C-terminal tail. In the active BLAminus structures, the tail is mostly coil and reaches up
to strands B2 and B3 of the N-terminal domain. In the “SRC-inactive” BLBplus state, the tail contains a
helix (residues 993-1002) in contact with the C-terminal domain, and then turns around with the C-terminus in contact with the I-helix of the C-terminal domain.</p>Download in Pymol:
<ul>
<li><a href="{{ url_for('static',filename='img/Fig_B1_EGFR_BLAminus_orange_tail.pse.zip') }}" target="_blank">EGFR_BLAminus</a><br></li>
<li><a href="{{ url_for('static',filename='img/Fig_B2_EGFR_BLBplus_blue_tail.pse.zip') }}" target="_blank">EGFR_BLBplus</a></li>
</ul>

<p style="text-align:justify;"><b>C</b>. Results of searches with the Kincore website<br> ATP-bound BLAminus structures from 14
different kinases in 6 kinase families (left) and BBAminus structures from 29 kinases with bound
Type 2 inhibitors (right).In the
BLAminus structures, the position of the DFG Phe (in yellow) and the conformation of the DFGmotif at
the beginning of the activation loop (shown in magenta) and the overall position of the activation loop
are consistent across the structures.</p>
Download in Pymol:
<ul>
<li><a href="{{ url_for('static',filename='img/Fig_C1_BLAminus with ATP.pse.zip') }}" target="_blank">BLAminus_ATP</a><br></li>
<li><a href="{{ url_for('static',filename='img/Fig_C2_TYR_BBAminus_Type2.pse.zip') }}" target="_blank">BBAminus_Type2</a></li>
</ul>

</div>

<div class="col-lg-6 col-md-12 col-sm-12 col-xs-12">
  <img src="{{ url_for('static', filename="img/Kincore-states-comparison.png") }}" width=100%>
</div>
</div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
              </div>
            </div>
          </div>

          <div class="row">
            <div class="col-lg-11 col-md-11 col-sm-11 col-xs-11">
              <div class="container-fluid very-dark-text">
                <div class="card mb-3">
                  <div class="card-header light-pink-background">
                    <h5>Protein kinase inhibitors</h5>
                  </div>


                  <div class="no-gutters">
                      <div class="card-body">
                        <div class="row">
                          <div class="col-lg-12">

                            <p>We have classified PK inhibitors into five groups based on the region of protein they bind to.</p>
                            <ul>
                              <li>Type 1</li>
                              <li>Type 1.5 - subdivided as Type 1.5_Front and Type 1.5_Back</li>
                              <li>Type 2</li>
                              <li>Type 3</li>
                              <li>Allosteric</li>
                            </ul>
                            <p>A list of FDA approved PK inhibitors with known structures can be accessed <a href="{{ url_for('FDA') }}" target="blank">here.</a></p>
                          </div>
                        </div>
                      </div>
                    </div>

                    <div class="card-footer text-muted">
                      <p>Defining a new nomenclature for the structures of active and inactive kinases. Vivek Modi and Roland L. Dunbrack, <em>PNAS</em> 2019, <strong>116(14) </strong>6818-6827.<br>
                      <a href="https://doi.org/10.1073/pnas.1814279116" target="_blank">Link to article</a></p>
                      <p>Structurally-Validated Multiple Sequence Alignment of 497 Human Protein Kinase Domains. Vivek Modi and Roland L. Dunbrack, <em>Sci Rep</em> 2019, <strong>9, 19790</strong>.<br>
                      <a href="https://www.nature.com/articles/s41598-019-56499-4" target="_blank">Link to article</a></p>
                    </div>
                  </div>
              </div>
            </div>
          </div>



    {% endblock %}
  </body>

</html>
